packages/make-hardener/src/main.js

```diff
@@ -33,113 +59,216 @@ const { ownKeys } = Reflect;
-          }
-        });
+  const patchReflectMethods = {
+    // Note that Object.preventExtensions has a slightly different
+    // semantics than Reflect.preventExtensions
+    preventExtensions(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
       }
+      return originalReflectPreventExtensions(target);
+    },
+  };
+
+  if (ENABLE_ONLY_NON_EXTENSIBLE) {
+    defineProperties(
+      Object,
+      Object.getOwnPropertyDescriptors(patchObjectMethods),
+    );
+
+    defineProperties(
+      Reflect,
+      Object.getOwnPropertyDescriptors(patchReflectMethods),
+    );
+  }
+
+  /**
+   * if ENABLE_ONLY_NON_EXTENSIBLE then this will be the
+   * patched one. Otherwise it will be the original.
+   */
+  const { freeze: patchedObjectFreeze } = Object;
 
-      function dequeue() {
-        // New values added before forEach() has finished will be visited.
-        toFreeze.forEach(freezeAndTraverse); // todo curried forEach
+  /**
+   * Currently not exported. But we should consider exporting it
+   * as a SES peer of isFrozen, etc. Of course, we'd need to remove
+   * the `path` parameter first.
+   *
+   * @param {any} val
+   * @param {string=} path
+   * @returns {boolean}
+   */
+  const isHardened = (val, path) => {
+    if (val !== Object(val)) {
+      return true;
+    }
+    let answer1;
+    let answer2;
+    if (ENABLE_OLD_HARDENED) {
+      answer1 = oldHardened.has(val);
+    }
+    if (ENABLE_ONLY_NON_EXTENSIBLE) {
+      // We check isFrozen, but we accumulate objects when they become
+      // non-extensible, because once non-extensible they might become frozen,
+      // and we can more feasibly catch all operations that might make them
+      // non-extensible.
+      answer2 = isFrozen(val) && !onlyNonExtensible.has(val);
+    }
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      if (answer1 !== answer2) {
+        // Not necessarily an error. Might be one of the three safe
+        // exceptions explained in the README.md. We should test these
+        // outliers, because if it is anything else, we may have a
+        // serious bug.
+        console.log(`HARDEN disagreement: ${answer1} vs ${answer2} at ${path}`);
       }
+    }
 
-      function commit() {
-        // todo curried forEach
-        // we capture the real WeakSet.prototype.add above, in case someone
-        // changes it. The two-argument form of forEach passes the second
-        // argument as the 'this' binding, so we add to the correct set.
-        toFreeze.forEach(hardened.add, hardened);
+    return !!(answer1 || answer2);
+  };
+
+  const moreObjectsHardened = toHarden => {
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      // TODO What consistency check should we put here?
+    }
+    if (ENABLE_OLD_HARDENED) {
+      toHarden.forEach(weakSetAdd, oldHardened);
+    }
+    if (ENABLE_ONLY_NON_EXTENSIBLE) {
+      toHarden.forEach(weakSetDelete, onlyNonExtensible);
```

Comment:
We should only need to delete from the set things that were already nonExtensible?

Reply:
I do not understand.

Reply:
@dtribble I don't understand this comment.